### ISSUE TITLE
"duplicates found in archive for WAL segment" with different timelines.

### ISSUE THREAD

**Issue by blogh** [pgbackrest/pgbackrest#420]
Hi,

Today, I was contacted by a team in my company.
They restored a backup with `pgbackrest restore --stanza=mtdi01 --type=immediate --delta --link-all --process-max=2`.

```
The history.
    full backup: 20171005-152258F
        timestamp start/stop: 2017-10-05 15:22:58 / 2017-10-05 15:28:14
        wal start/stop: 00000006000000C700000021 / 00000006000000C700000021
        database size: 829.2MB, backup size: 829.2MB
        repository size: 111.9MB, repository backup size: 111.9MB

    incr backup: 20171005-152258F_20171005-202804I
        timestamp start/stop: 2017-10-05 20:28:04 / 2017-10-05 20:37:25
        wal start/stop: 00000007000000C700000060 / 00000007000000C700000066
        database size: 1.2GB, backup size: 568MB
        repository size: 152.9MB, repository backup size: 73.1MB
        backup reference list: 20171005-152258F
```

		
After restoring they got this, when they did a backup `pgbackrest backup --stanza=mtdi01 --type=full --start-fast`:
```
ERROR [045]: : duplicates found in archive for WAL segment XXXXXXXX000000C70000002E: 00000006000000C70000002E-a00e1030c12fd58e7fc56acafece576e9e29e538.gz, 00000007000000C70000002E-ed2089a928b6916f16c8ef382e103a1264581990.gz, 00000008000000C70000002E-b8c1191b8d97a37d3b3c5042310008fcbcb36863.gz
2017-10-06 13:25:57.795 P00   INFO: backup command end: aborted with exception [045]
```

DEBUG:
```
2017-10-06 13:33:02.830 P00   INFO: backup stop archive = 00000008000000C700000032, lsn = C7/320000B8
2017-10-06 13:33:02.830 P00  DEBUG:     Protocol::Helper::protocolDestroy(): bComplete = true, iRemoteIdx = [undef], strRemoteType = [undef]
2017-10-06 13:33:02.830 P00  DEBUG:     Protocol::Helper::protocolDestroy=>: iExitStatus = 0
2017-10-06 13:33:02.830 P00  DEBUG:     Storage::Local->exists(): strFileExp = /home/postgres/backup/pgbackrest1.0/backup/mtdi01/20171006-132508F/backup.manifest
2017-10-06 13:33:02.830 P00  DEBUG:     Storage::Local->exists=>: bExists = false
2017-10-06 13:33:03.433 P00  DEBUG:     Storage::Local->openWrite(): bAtomic = <false>, bPathCreate = <false>, lTimestamp = [undef], rhyFilter = [undef], strGroup = [undef], strMode = <0640>, strUser = [undef], xFileExp = /home/postgres/backup/pgbackrest1.0/backup/mtdi01/20171006-132508F/backup.manifest.copy
2017-10-06 13:33:03.500 P00  DEBUG:     Backup::Backup->process: retrieve archive logs 00000008000000C700000032:00000008000000C700000032
2017-10-06 13:33:03.501 P00  DEBUG:     Archive::Info->new(): bIgnoreMissing = <false>, bLoad = <true>, bRequired = true, strArchiveClusterPath = /home/postgres/backup/pgbackrest1.0/archive/mtdi01
2017-10-06 13:33:03.501 P00  DEBUG:     Storage::Local->openRead(): bIgnoreMissing = true, rhyFilter = [undef], xFileExp = /home/postgres/backup/pgbackrest1.0/archive/mtdi01/archive.info
2017-10-06 13:33:03.502 P00  DEBUG:     Archive::Info->archiveId(): strDbVersion = [undef], ullDbSysId = [undef]
2017-10-06 13:33:03.502 P00  DEBUG:     Archive::Info->archiveId=>: strArchiveId = 9.4-1
2017-10-06 13:33:03.502 P00  DEBUG:     Archive::Common::lsnFileRange(): $strDbVersion = 9.4, strLsnStart = C7/32000028, strLsnStop = C7/320000B8
2017-10-06 13:33:03.503 P00  DEBUG:     Archive::Common::lsnFileRange=>: stryWalFileName = (000000C700000032)
2017-10-06 13:33:03.503 P00  DEBUG:     Archive::Common::walSegmentFind(): iWaitSeconds = 1, oStorageRepo = [object], strArchiveId = 9.4-1, strWalSegment = 000000C700000032
2017-10-06 13:33:03.503 P00  DEBUG:     Storage::Local->list(): bIgnoreMissing = true, strExpression = [0-F]{8}000000C7, strPathExp = <REPO:ARCHIVE>/9.4-1, strSortOrder = <forward>
2017-10-06 13:33:03.503 P00  DEBUG:     Storage::Local->list=>: stryFileList = (00000006000000C7, 00000007000000C7, 00000008000000C7)
2017-10-06 13:33:03.503 P00  DEBUG:     Storage::Local->list(): bIgnoreMissing = true, strExpression = ^00000006000000C700000032-[0-f]{40}(\.gz){0,1}$, strPathExp = <REPO:ARCHIVE>/9.4-1/00000006000000C7, strSortOrder = <forward>
2017-10-06 13:33:03.504 P00  DEBUG:     Storage::Local->list=>: stryFileList = (00000006000000C700000032-09fb955314e327620c448006b86767640b755d00.gz)
2017-10-06 13:33:03.504 P00  DEBUG:     Storage::Local->list(): bIgnoreMissing = true, strExpression = ^00000007000000C700000032-[0-f]{40}(\.gz){0,1}$, strPathExp = <REPO:ARCHIVE>/9.4-1/00000007000000C7, strSortOrder = <forward>
2017-10-06 13:33:03.504 P00  DEBUG:     Storage::Local->list=>: stryFileList = (00000007000000C700000032-1147a3851c983526a5fc66a90b8427a3220c1743.gz)
2017-10-06 13:33:03.505 P00  DEBUG:     Storage::Local->list(): bIgnoreMissing = true, strExpression = ^00000008000000C700000032-[0-f]{40}(\.gz){0,1}$, strPathExp = <REPO:ARCHIVE>/9.4-1/00000008000000C7, strSortOrder = <forward>
2017-10-06 13:33:03.505 P00  DEBUG:     Storage::Local->list=>: stryFileList = (00000008000000C700000032-a09c9789feef08c91d138eca9f38752d5a005a8b.gz)
2017-10-06 13:33:03.505 P00  ERROR: [045]: duplicates found in archive for WAL segment XXXXXXXX000000C700000032: 00000006000000C700000032-09fb955314e327620c448006b86767640b755d00.gz, 00000007000000C700000032-1147a3851c983526a5fc66a90b8427a3220c1743.gz, 00000008000000C700000032-a09c9789feef08c91d138eca9f38752d5a005a8b.gz
 at /usr/share/perl5/pgBackRest/Archive/Common.pm line 330.
                                        pgBackRest::Archive::Common::walSegmentFind('pgBackRest::Storage::Local=HASH(0x2f8ef70)', '9.4-1', '000000C700000032', 1) called at /usr/share/perl5/pgBackRest/Backup/Backup.pm line 890
                                        pgBackRest::Backup::Backup::process('pgBackRest::Backup::Backup=HASH(0x2f02360)') called at /usr/bin/pgbackrest line 245
                                        eval {...} called at /usr/bin/pgbackrest line 275

                                   2017-10-06 13:33:03.536 P00  DEBUG:     Common::Exit::exitSafe(): iExitCode = [undef], oException = [object], strSignal = [undef]
2017-10-06 13:33:03.536 P00  DEBUG:     Protocol::Helper::protocolDestroy(): bComplete = false, iRemoteIdx = [undef], strRemoteType = [undef]
2017-10-06 13:33:03.536 P00  DEBUG:     Protocol::Helper::protocolDestroy=>: iExitStatus = 0
2017-10-06 13:33:03.536 P00  DEBUG:     Common::Lock::lockRelease(): bFailOnNoLock = false
2017-10-06 13:33:03.536 P00  ERROR: [045]: duplicates found in archive for WAL segment XXXXXXXX000000C700000032: 00000006000000C700000032-09fb955314e327620c448006b86767640b755d00.gz, 00000007000000C700000032-1147a3851c983526a5fc66a90b8427a3220c1743.gz, 00000008000000C700000032-a09c9789feef08c91d138eca9f38752d5a005a8b.gz
 at /usr/share/perl5/pgBackRest/Common/Log.pm line 557.
                                        pgBackRest::Common::Log::logException('pgBackRest::Common::Exception=HASH(0x30138e8)') called at /usr/share/perl5/pgBackRest/Common/Exit.pm line 79
                                        pgBackRest::Common::Exit::exitSafe(undef, '') called at /usr/bin/pgbackrest line 276

                                   2017-10-06 13:33:03.536 P00   INFO: backup command end: aborted with exception [045]
2017-10-06 13:33:03.536 P00  DEBUG:     Common::Exit::exitSafe=>: iExitCode = 45
```

When I check in PostgreSQL's log, the archive push works.

I dont quite understand why the wal in other timelines are seen as a problem.
In "walSegmentFind", I see two ways of dealing with the wals depending of if the timeline is found or not. I haven't checked further yet. Maybe I am missing something obvious ?

I will set `--no-archive-check` (without --archive-copy),  it will fix (/hide) the problem for now. But I need to understand this. Do you have an idea please ?

Benoit

EDIT: --archive-check => --no-archive-check & command for backup

**Comment by dwsteele** [comment]
If there are two masters archiving into the same repo then the results of a restore can be unpredictable.  What if you restore a backup from timeline 5 with `recovery_target_timeline=latest`?  It will switch where the new master diverged and play along that master's timeline which is probably not what you expect.  Worse, the default for restores is to pick the most recent backup which might be from the other master.

In short - it's not OK to have two masters archiving and backing up into the same stanza.  There are just too many ways it can go wrong.  We already have plans to improve the error message with some hints to help diagnose the issue.

So why is it possible to archive at all?  We determined it was better to allow archiving to continue rather than stop it completely and risk a disk on the master filling up and causing a Postgres panic.  Also, there are times (during promotions) when there can be a brief timeline overlap (one WAL segment), but there should never be overlap during a backup.

You should create a new stanza for the new master and remove timeline 6 from the archive in the old repo.  If you set `--no-archive-check` already you may also want to delete the backups taken on timeline 6.

**Comment by blogh** [comment]
First, Thanks a lot for the quick answer.

English is not my first language so please allow me to summarize what they did and rephrase what you explained.

From what I gathered they did: Backup, Tests, Restore on the same PostgreSQL Cluster (which I didn't say clearly before) several times. (in that order with different tests).
So they have something like this (retention is 3 backups on disk)
- Backup 1 in tl 6 => Tests1 => restore => tl switch to 7
- Backup 2 in tl 7 => Tests2 => restore => tl switch to 8
- Backup 3 in tl 8 =>  error


From your answer, I understand that:
-  In general: I should create a new stanza after a each restore even if it's on the same cluster.
-  To solve this: I should delete wals on tl6 (expiring the backup 1 should do the trick ? (I will test on monday, I dont have remote acces outside of work))

Thanks
Benoit

**Comment by dwsteele** [comment]
Thanks for the additional information.  I have seen this happen when two masters are running on the same stanza and I jumped to the wrong conclusion.

It looks like your use case is valid and pgBackRest is not behaving correctly.  This should be a WARN and not an ERROR since the timestamps on the other timelines are older than the current timeline.

The best you can do for now is to disable `archive-check` as you have done.  I would not recommend doing this on a production system but it sounds like this database is running in a lower environment.

In any case, I have promoted this issue to a bug and we will produce a fix by the next release.

**Comment by blogh** [comment]
Thanks !

**Comment by dwsteele** [comment]
Fixed at c94e52f697cdae6.
